include/revng/Model/Processing.h

```diff
@@ -23,4 +23,99 @@ unsigned
 dropTypesDependingOnDefinitions(TupleTree<model::Binary> &Binary,
                                 const std::set<const TypeDefinition *> &Defs);
 
+namespace detail {
+
+template<typename PreVisitorType, typename PostVisitorType>
+struct TypeVisitorImpl {
+  PreVisitorType &&PreVisitor;
+  PostVisitorType &&PostVisitor;
+
+  RecursiveCoroutine<> visit(model::UpcastableTypeDefinition &Definition) {
```

Comment:
Are you sure we can't do this with `edges()`? Similar to what `createTypeGraph` does.

I'm trying quite hard to avoid enumerating in many places all the things that you can have in the model, it's easy for them to go out of sync.

Reply:
>     // TODO: consider merging the two (by providing `owningEdges()` that both
>     //       this and `edges()` can reuse).

How difficult is this to do?
Let's strive not to enumerate model parts in centralized places if we can avoid it.

> Don't forget that we're removing types here.

Not sure what you mean. :thinking:

Reply:
(closing this, we discussed it further in a different thread)